File: week02/compare/memory.c
/*
 * memory.c - Demonstrating stack vs heap allocation in C
 *
 * Key concepts:
 * - Stack: automatic storage, fast, limited size, scoped lifetime
 * - Heap: manual management, slower, large capacity, explicit lifetime
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ============================================================
 * SECTION 1: Stack Allocation
 * ============================================================
 * Stack-allocated data lives in the function's stack frame.
 * It is automatically deallocated when the function returns.
 */

void stack_example(void) {
    /* Stack-allocated array - fixed size, known at compile time */
    char greeting[32] = "Hello, World!";

    printf("Stack string: %s\n", greeting);
    printf("  Address: %p\n", (void *)greeting);
    printf("  sizeof:  %zu bytes (full buffer)\n", sizeof(greeting));
    printf("  strlen:  %zu bytes (actual content)\n\n", strlen(greeting));

    /* We can modify it within bounds */
    strcpy(greeting, "Modified!");
    printf("After modification: %s\n\n", greeting);

    /* greeting is automatically freed when this function returns */
}

/* ============================================================
 * SECTION 2: The Dangling Pointer Problem
 * ============================================================
 * DANGER: Returning a pointer to stack memory is undefined behavior!
 * The memory is deallocated when the function returns.
 */

/* WARNING: This function has a bug - DO NOT do this! */
char *dangerous_stack_return(void) {
    char local_buffer[64] = "I live on the stack";

    /* BUG: returning pointer to local variable! */
    /* The memory will be invalid after this function returns */
    return local_buffer;  /* Compiler may warn about this */
}

void demonstrate_dangling_pointer(void) {
    printf("=== Dangling Pointer Demonstration ===\n");

    char *dangerous = dangerous_stack_return();

    /* This is UNDEFINED BEHAVIOR - the memory may be:
     * - Overwritten by subsequent function calls
     * - Appear to "work" sometimes (making bugs hard to find)
     * - Crash the program
     * - Contain garbage data
     */
    printf("Dangerous pointer contains: %s\n", dangerous);
    printf("(This may appear to work, crash, or show garbage)\n\n");
}

/* ============================================================
 * SECTION 3: Heap Allocation
 * ============================================================
 * Heap memory persists until explicitly freed.
 * The programmer is responsible for:
 * - Allocating (malloc, calloc, realloc)
 * - Freeing (free)
 * - Not using after free
 * - Not freeing twice
 */

char *safe_heap_return(const char *input) {
    /* Allocate on heap - memory persists after function returns */
    size_t len = strlen(input) + 1;  /* +1 for null terminator */
    char *heap_string = malloc(len);

    if (heap_string == NULL) {
        fprintf(stderr, "malloc failed!\n");
        return NULL;
    }

    strcpy(heap_string, input);

    /* Safe to return - caller owns this memory now */
    /* But who frees it? The caller must remember to! */
    return heap_string;
}

void heap_example(void) {
    printf("=== Heap Allocation Example ===\n");

    /* Caller receives ownership of heap memory */
    char *my_string = safe_heap_return("Hello from the heap!");

    if (my_string != NULL) {
        printf("Heap string: %s\n", my_string);
        printf("  Address: %p\n\n", (void *)my_string);

        /* We can resize heap allocations */
        my_string = realloc(my_string, 100);
        strcat(my_string, " Plus more text!");
        printf("After realloc: %s\n\n", my_string);

        /* CRITICAL: We must free the memory */
        free(my_string);

        /* After free, my_string is a "dangling pointer" */
        /* Using it is undefined behavior */
        /* Good practice: set to NULL after free */
        my_string = NULL;
    }
}

/* ============================================================
 * SECTION 4: Common Memory Bugs in C
 * ============================================================
 */

void memory_leak_example(void) {
    printf("=== Memory Leak Example ===\n");

    char *leaked = malloc(100);
    strcpy(leaked, "I will be leaked!");

    /* Oops - we return without freeing! */
    /* This memory is now unreachable and cannot be reclaimed */
    /* In long-running programs, leaks accumulate and exhaust memory */

    printf("Allocated but never freed: %s\n", leaked);
    printf("(Memory leak - no way to recover this memory)\n\n");

    /* Fix: free(leaked); */
}

void double_free_example(void) {
    printf("=== Double Free Example (COMMENTED OUT - would crash) ===\n");

    char *data = malloc(50);
    strcpy(data, "Free me once");

    free(data);

    /* DANGER: Uncommenting this would cause undefined behavior!
     * Double-free can:
     * - Crash the program
     * - Corrupt the heap
     * - Create security vulnerabilities
     */
    /* free(data); */  /* DON'T DO THIS! */

    printf("First free: OK\n");
    printf("Second free: Would crash or corrupt memory!\n\n");
}

void use_after_free_example(void) {
    printf("=== Use After Free Example ===\n");

    char *data = malloc(50);
    strcpy(data, "Valid data");
    printf("Before free: %s\n", data);

    free(data);

    /* DANGER: data now points to freed memory!
     * Reading or writing is undefined behavior.
     * May appear to work, making bugs hard to find.
     */

    /* This is a bug - the memory has been freed! */
    /* printf("After free: %s\n", data); */  /* DON'T DO THIS! */

    printf("After free: (accessing would be undefined behavior)\n\n");
}

/* ============================================================
 * SECTION 5: Passing Strings to Functions
 * ============================================================
 * Who owns what? C doesn't enforce this - it's all by convention.
 */

/* This function borrows the string - does not modify or free */
void print_string(const char *str) {
    printf("Borrowed string: %s\n", str);
    /* We don't free str - we're just borrowing it */
}

/* This function takes ownership - caller should not use after */
void take_ownership(char *str) {
    printf("Taking ownership of: %s\n", str);
    /* We own it now, we must free it */
    free(str);
}

/* This function modifies the borrowed string */
void modify_string(char *str, size_t buffer_size) {
    /* We modify but don't own - caller still responsible for memory */
    strncat(str, " [modified]", buffer_size - strlen(str) - 1);
}

void ownership_confusion_demo(void) {
    printf("=== Ownership Confusion Demo ===\n");

    char *heap_str = safe_heap_return("Heap string");

    /* Borrowing is OK */
    print_string(heap_str);

    /* After this call, heap_str is freed but still points to memory! */
    take_ownership(heap_str);

    /* BUG: heap_str is now dangling - don't use it! */
    /* print_string(heap_str); */  /* Would be undefined behavior! */

    printf("(heap_str is now invalid - using it would be a bug)\n\n");
}

/* ============================================================
 * MAIN
 * ============================================================
 */

int main(void) {
    printf("========================================\n");
    printf("C Memory Management Demonstration\n");
    printf("========================================\n\n");

    printf("=== Stack Allocation ===\n");
    stack_example();

    printf("\n");
    demonstrate_dangling_pointer();

    heap_example();

    memory_leak_example();

    double_free_example();

    use_after_free_example();

    ownership_confusion_demo();

    printf("========================================\n");
    printf("Summary of C Memory Pitfalls:\n");
    printf("========================================\n");
    printf("1. Dangling pointers (returning stack memory)\n");
    printf("2. Memory leaks (forgetting to free)\n");
    printf("3. Double free (freeing twice)\n");
    printf("4. Use after free (accessing freed memory)\n");
    printf("5. Ownership confusion (who frees what?)\n");
    printf("6. Buffer overflows (writing past allocation)\n");
    printf("\nRust prevents ALL of these at compile time!\n");

    return 0;
}

File: week02/compare/memory.rs
/*
 * memory.rs - Demonstrating how Rust prevents C's memory pitfalls
 *
 * Key concepts:
 * - Ownership: Each value has exactly one owner
 * - Borrowing: References let you access data without taking ownership
 * - Lifetimes: The compiler tracks how long references are valid
 * - Drop: Automatic deallocation when owner goes out of scope
 */

/* ============================================================
 * SECTION 1: Stack Allocation (similar to C)
 * ============================================================
 * Stack allocation works similarly, but Rust tracks ownership.
 */

fn stack_example() {
    // Stack-allocated array - fixed size
    let greeting: [u8; 32] = *b"Hello, World!\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0";

    // More commonly, we use &str for string literals (stored in binary)
    let message: &str = "Hello, World!";

    println!("Stack/static string: {}", message);
    println!("  Length: {} bytes", message.len());

    // greeting goes out of scope here - automatically cleaned up
    // No manual free needed, no way to forget!
    let _ = greeting; // silence unused warning
}

/* ============================================================
 * SECTION 2: Rust PREVENTS Dangling Pointers
 * ============================================================
 * This code would NOT COMPILE in Rust - the compiler catches the bug!
 */

// UNCOMMENT TO SEE COMPILER ERROR:
// fn dangerous_stack_return() -> &str {
//     let local_string = String::from("I live on the stack");
//
//     // ERROR: cannot return reference to local variable
//     // The compiler knows local_string will be dropped!
//     &local_string
// }

// The SAFE way: return an owned value (moves ownership to caller)
fn safe_return() -> String {
    let local_string = String::from("I live on the heap");

    // Ownership is MOVED to the caller
    // local_string is not dropped here - caller now owns it
    local_string
}

fn demonstrate_no_dangling_pointer() {
    println!("=== No Dangling Pointers in Rust ===");

    // We receive ownership of the String
    let my_string = safe_return();

    println!("Received owned string: {}", my_string);
    println!("This is always safe - no undefined behavior possible!\n");

    // my_string is automatically dropped (freed) here
}

/* ============================================================
 * SECTION 3: Heap Allocation with String
 * ============================================================
 * String is Rust's heap-allocated, growable string type.
 * Ownership is explicit and enforced by the compiler.
 */

fn heap_example() {
    println!("=== Heap Allocation Example ===");

    // String allocates on the heap
    let mut my_string = String::from("Hello from the heap!");

    println!("Heap string: {}", my_string);
    println!("  Length:   {} bytes", my_string.len());
    println!("  Capacity: {} bytes\n", my_string.capacity());

    // We can grow it (like realloc, but automatic)
    my_string.push_str(" Plus more text!");
    println!("After push: {}", my_string);
    println!("  New capacity: {} bytes\n", my_string.capacity());

    // NO MANUAL FREE NEEDED!
    // my_string is automatically freed when it goes out of scope
    // This is called "Drop" in Rust
}

/* ============================================================
 * SECTION 4: Rust PREVENTS Common Memory Bugs
 * ============================================================
 */

fn no_memory_leaks() {
    println!("=== No Memory Leaks (normally) ===");

    let data = String::from("I will be automatically freed!");

    println!("Allocated: {}", data);

    // When data goes out of scope, it's automatically dropped
    // No way to "forget" to free it!
    // (Note: Rust does have mem::forget for special cases, but you have to explicitly use it)

    println!("(data will be freed automatically at end of scope)\n");
}

fn no_double_free() {
    println!("=== No Double Free ===");

    let data = String::from("Free me once");

    // This MOVES ownership - data can no longer be used
    let _moved = data;

    // UNCOMMENT TO SEE COMPILER ERROR:
    // let _double = data;  // ERROR: use of moved value

    println!("Ownership moved - original variable is now invalid");
    println!("Compiler prevents any possibility of double free!\n");

    // Only _moved will be dropped (once!)
}

fn no_use_after_free() {
    println!("=== No Use After Free ===");

    let data = String::from("Valid data");
    println!("Before drop: {}", data);

    // Explicitly drop (free) the data
    drop(data);

    // UNCOMMENT TO SEE COMPILER ERROR:
    // println!("After drop: {}", data);  // ERROR: use of moved value

    println!("(accessing data after drop is a compile-time error!)\n");
}

/* ============================================================
 * SECTION 5: Borrowing - Safe References
 * ============================================================
 * Borrowing lets you access data without taking ownership.
 * The compiler ensures references are always valid.
 */

// Borrows the string (immutable reference) - cannot modify
fn print_string(s: &str) {
    println!("Borrowed string: {}", s);
    // We don't own s - it will NOT be dropped when this function returns
}

// Takes OWNERSHIP - caller loses access
fn take_ownership(s: String) {
    println!("Taking ownership of: {}", s);
    // s is dropped at end of function - memory freed
}

// Borrows mutably - can modify, but doesn't own
fn modify_string(s: &mut String) {
    s.push_str(" [modified]");
    // We borrowed mutably, but don't own - not dropped here
}

fn clear_ownership_demo() {
    println!("=== Clear Ownership Demo ===");

    let mut heap_str = String::from("Heap string");

    // Borrowing is explicit with &
    print_string(&heap_str);

    // We still own heap_str, can use it again
    println!("Still valid: {}", heap_str);

    // Mutable borrow
    modify_string(&mut heap_str);
    println!("After modification: {}", heap_str);

    // This MOVES ownership - heap_str is now invalid
    take_ownership(heap_str);

    // UNCOMMENT TO SEE COMPILER ERROR:
    // println!("{}", heap_str);  // ERROR: use of moved value

    println!("(heap_str is now invalid - compiler enforces this!)\n");
}

/* ============================================================
 * SECTION 6: Lifetimes - Compiler Tracks Reference Validity
 * ============================================================
 */

// The lifetime 'a says: the returned reference lives as long as the input
fn longest<'a>(s1: &'a str, s2: &'a str) -> &'a str {
    if s1.len() > s2.len() {
        s1
    } else {
        s2
    }
}

fn lifetime_demo() {
    println!("=== Lifetime Demo ===");

    let string1 = String::from("hello");

    {
        let string2 = String::from("world!");
        let result = longest(&string1, &string2);
        println!("Longest: {}", result);
        // result is valid here because both string1 and string2 are valid
    }

    // UNCOMMENT TO SEE COMPILER ERROR:
    // This would be an error because string2 is dropped:
    // let string2 = String::from("world!");
    // let result;
    // {
    //     let string2 = String::from("world!");
    //     result = longest(&string1, &string2);
    // }  // string2 dropped here
    // println!("{}", result);  // ERROR: string2 does not live long enough

    println!("Compiler ensures references never outlive their data!\n");
}

/* ============================================================
 * SECTION 7: Comparison with C
 * ============================================================
 */

fn comparison_with_c() {
    println!("=== Rust vs C Memory Safety ===\n");

    println!("| C Problem              | Rust Solution                    |");
    println!("|------------------------|----------------------------------|");
    println!("| Dangling pointers      | Lifetime checking at compile time|");
    println!("| Memory leaks           | Automatic Drop when out of scope |");
    println!("| Double free            | Ownership - only one owner       |");
    println!("| Use after free         | Move semantics invalidate source |");
    println!("| Ownership confusion    | Explicit ownership in type system|");
    println!("| Buffer overflow        | Bounds checking (can be bypassed)|");
    println!();

    println!("Key insight: Rust's ownership rules are enforced at COMPILE TIME.");
    println!("There is NO runtime overhead for memory safety checks.");
    println!("You get C-like performance with memory safety guarantees.\n");
}

/* ============================================================
 * MAIN
 * ============================================================
 */

fn main() {
    println!("========================================");
    println!("Rust Memory Safety Demonstration");
    println!("========================================\n");

    println!("=== Stack Allocation ===");
    stack_example();
    println!();

    demonstrate_no_dangling_pointer();

    heap_example();

    no_memory_leaks();

    no_double_free();

    no_use_after_free();

    clear_ownership_demo();

    lifetime_demo();

    comparison_with_c();

    println!("========================================");
    println!("Rust Memory Safety Summary:");
    println!("========================================");
    println!("1. Ownership: Each value has exactly one owner");
    println!("2. Borrowing: & for read-only, &mut for read-write");
    println!("3. Lifetimes: Compiler tracks reference validity");
    println!("4. Drop: Automatic cleanup when owner goes out of scope");
    println!("5. No null: Option<T> makes absence explicit");
    println!();
    println!("All enforced at compile time = zero runtime cost!");
}
